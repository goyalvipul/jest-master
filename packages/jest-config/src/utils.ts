/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import {Config} from '@jest/types';
import {ValidationError} from 'jest-validate';
import Resolver from 'jest-resolve';
import chalk from 'chalk';

type ResolveOptions = {
  rootDir: Config.Path;
  key: string;
  filePath: Config.Path;
  optional?: boolean;
};

export const BULLET: string = chalk.bold('\u25cf ');
export const DOCUMENTATION_NOTE = `  ${chalk.bold(
  'Configuration Documentation:',
)}
  https://jestjs.io/docs/configuration.html
`;

const createValidationError = (message: string) =>
  new ValidationError(`${BULLET}Validation Error`, message, DOCUMENTATION_NOTE);

export const resolve = (
  resolver: string | null | undefined,
  {key, filePath, rootDir, optional}: ResolveOptions,
) => {
  const module = Resolver.findNodeModule(
    replaceRootDirInPath(rootDir, filePath),
    {
      basedir: rootDir,
      resolver: resolver || undefined,
    },
  );

  if (!module && !optional) {
    throw createValidationError(
      `  Module ${chalk.bold(filePath)} in the ${chalk.bold(
        key,
      )} option was not found.
         ${chalk.bold('<rootDir>')} is: ${rootDir}`,
    );
  }
  /// can cast as string since nulls will be thrown
  return module as string;
};

export const escapeGlobCharacters = (path: Config.Path): Config.Glob =>
  path.replace(/([()*{}\[\]!?\\])/g, '\\$1');

export const replaceRootDirInPath = (
  rootDir: Config.Path,
  filePath: Config.Path,
): string => {
  if (!/^<rootDir>/.test(filePath)) {
    return filePath;
  }

  return path.resolve(
    rootDir,
    path.normalize('./' + filePath.substr('<rootDir>'.length)),
  );
};

const _replaceRootDirInObject = <T extends ReplaceRootDirConfigObj>(
  rootDir: Config.Path,
  config: T,
): T => {
  const newConfig = {} as T;
  for (const configKey in config) {
    newConfig[configKey] =
      configKey === 'rootDir'
        ? config[configKey]
        : _replaceRootDirTags(rootDir, config[configKey]);
  }
  return newConfig;
};

type OrArray<T> = T | Array<T>;
type ReplaceRootDirConfigObj = Record<string, Config.Path>;
type ReplaceRootDirConfigValues =
  | OrArray<ReplaceRootDirConfigObj>
  | OrArray<RegExp>
  | OrArray<Config.Path>;

export const _replaceRootDirTags = <T extends ReplaceRootDirConfigValues>(
  rootDir: Config.Path,
  config: T,
): T => {
  if (config == null) {
    return config;
  }
  switch (typeof config) {
    case 'object':
      if (Array.isArray(config)) {
        /// can be string[] or {}[]
        return config.map(item => _replaceRootDirTags(rootDir, item)) as T;
      }
      if (config instanceof RegExp) {
        return config;
      }

      return _replaceRootDirInObject(
        rootDir,
        config as ReplaceRootDirConfigObj,
      ) as T;
    case 'string':
      return replaceRootDirInPath(rootDir, config) as T;
  }
  return config;
};

export const resolveWithPrefix = (
  resolver: string | undefined | null,
  {
    filePath,
    humanOptionName,
    optionName,
    prefix,
    rootDir,
  }: {
    filePath: string;
    humanOptionName: string;
    optionName: string;
    prefix: string;
    rootDir: Config.Path;
  },
) => {
  const fileName = replaceRootDirInPath(rootDir, filePath);
  let module = Resolver.findNodeModule(`${prefix}${fileName}`, {
    basedir: rootDir,
    resolver: resolver || undefined,
  });
  if (module) {
    return module;
  }

  try {
    return require.resolve(`${prefix}${fileName}`);
  } catch (e) {}

  module = Resolver.findNodeModule(fileName, {
    basedir: rootDir,
    resolver: resolver || undefined,
  });
  if (module) {
    return module;
  }

  try {
    return require.resolve(fileName);
  } catch (e) {}

  throw createValidationError(
    `  ${humanOptionName} ${chalk.bold(
      fileName,
    )} cannot be found. Make sure the ${chalk.bold(
      optionName,
    )} configuration option points to an existing node module.`,
  );
};

/**
 * Finds the test environment to use:
 *
 * 1. looks for jest-environment-<name> relative to project.
 * 1. looks for jest-environment-<name> relative to Jest.
 * 1. looks for <name> relative to project.
 * 1. looks for <name> relative to Jest.
 */
export const getTestEnvironment = ({
  rootDir,
  testEnvironment: filePath,
}: {
  rootDir: Config.Path;
  testEnvironment: string;
}) =>
  resolveWithPrefix(undefined, {
    filePath,
    humanOptionName: 'Test environment',
    optionName: 'testEnvironment',
    prefix: 'jest-environment-',
    rootDir,
  });

/**
 * Finds the watch plugins to use:
 *
 * 1. looks for jest-watch-<name> relative to project.
 * 1. looks for jest-watch-<name> relative to Jest.
 * 1. looks for <name> relative to project.
 * 1. looks for <name> relative to Jest.
 */
export const getWatchPlugin = (
  resolver: string | undefined | null,
  {filePath, rootDir}: {filePath: string; rootDir: Config.Path},
) =>
  resolveWithPrefix(resolver, {
    filePath,
    humanOptionName: 'Watch plugin',
    optionName: 'watchPlugins',
    prefix: 'jest-watch-',
    rootDir,
  });

/**
 * Finds the runner to use:
 *
 * 1. looks for jest-runner-<name> relative to project.
 * 1. looks for jest-runner-<name> relative to Jest.
 * 1. looks for <name> relative to project.
 * 1. looks for <name> relative to Jest.
 */
export const getRunner = (
  resolver: string | undefined | null,
  {filePath, rootDir}: {filePath: string; rootDir: Config.Path},
) =>
  resolveWithPrefix(resolver, {
    filePath,
    humanOptionName: 'Jest Runner',
    optionName: 'runner',
    prefix: 'jest-runner-',
    rootDir,
  });

type JSONString = string & {readonly $$type: never}; // newtype
export const isJSONString = (text?: JSONString | string): text is JSONString =>
  text != null &&
  typeof text === 'string' &&
  text.startsWith('{') &&
  text.endsWith('}');

export const getSequencer = (
  resolver: string | undefined | null,
  {filePath, rootDir}: {filePath: string; rootDir: Config.Path},
) =>
  resolveWithPrefix(resolver, {
    filePath,
    humanOptionName: 'Jest Sequencer',
    optionName: 'testSequencer',
    prefix: 'jest-sequencer-',
    rootDir,
  });
