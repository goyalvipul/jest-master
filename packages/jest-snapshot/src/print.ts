/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import diff, {getStringDiff} from 'jest-diff';
import getType, {isPrimitive} from 'jest-get-type';
import {
  EXPECTED_COLOR,
  RECEIVED_COLOR,
  getLabelPrinter,
  printDiffOrStringify,
} from 'jest-matcher-utils';
import prettyFormat from 'pretty-format';
import {unescape} from './utils';

const isLineDiffable = (received: any): boolean => {
  const receivedType = getType(received);

  if (isPrimitive(received)) {
    return typeof received === 'string' && received.includes('\n');
  }

  if (
    receivedType === 'date' ||
    receivedType === 'function' ||
    receivedType === 'regexp'
  ) {
    return false;
  }

  if (received instanceof Error) {
    return false;
  }

  if (
    receivedType === 'object' &&
    typeof (received as any).asymmetricMatch === 'function'
  ) {
    return false;
  }

  return true;
};

export const printDiffOrStringified = (
  expectedSerializedTrimmed: string,
  receivedSerializedTrimmed: string,
  received: unknown,
  expectedLabel: string,
  receivedLabel: string,
  expand: boolean, // CLI options: true if `--expand` or false if `--no-expand`
): string => {
  if (typeof received === 'string') {
    if (
      expectedSerializedTrimmed.length >= 2 &&
      expectedSerializedTrimmed.startsWith('"') &&
      expectedSerializedTrimmed.endsWith('"') &&
      receivedSerializedTrimmed === unescape(prettyFormat(received))
    ) {
      // The expected snapshot looks like a stringified string.
      // The received serialization is default stringified string.

      // Undo default serialization of expected snapshot:
      // Remove enclosing double quote marks.
      // Remove backslash escape preceding backslash here,
      // because unescape replaced it only preceding double quote mark.
      return printDiffOrStringify(
        expectedSerializedTrimmed.slice(1, -1).replace(/\\\\/g, '\\'),
        received,
        expectedLabel,
        receivedLabel,
        expand,
      );
    }

    // Display substring highlight even when strings have custom serialization.
    const result = getStringDiff(
      expectedSerializedTrimmed,
      receivedSerializedTrimmed,
      {
        aAnnotation: expectedLabel,
        bAnnotation: receivedLabel,
        expand,
      },
    );

    if (result !== null) {
      if (result.isMultiline) {
        return result.annotatedDiff;
      }

      // Because not default stringify, call EXPECTED_COLOR and RECEIVED_COLOR
      // This is reason to call getStringDiff instead of printDiffOrStringify
      // Because there is no closing double quote mark at end of single lines,
      // future improvement is to call replaceSpacesAtEnd if it becomes public.
      const printLabel = getLabelPrinter(expectedLabel, receivedLabel);
      return (
        printLabel(expectedLabel) +
        EXPECTED_COLOR(result.a) +
        '\n' +
        printLabel(receivedLabel) +
        RECEIVED_COLOR(result.b)
      );
    }
  }

  if (
    (expectedSerializedTrimmed.includes('\n') ||
      receivedSerializedTrimmed.includes('\n')) &&
    isLineDiffable(received)
  ) {
    return diff(expectedSerializedTrimmed, receivedSerializedTrimmed, {
      aAnnotation: expectedLabel,
      bAnnotation: receivedLabel,
      expand,
    }) as string;
  }

  const printLabel = getLabelPrinter(expectedLabel, receivedLabel);
  return (
    printLabel(expectedLabel) +
    EXPECTED_COLOR(expectedSerializedTrimmed) +
    '\n' +
    printLabel(receivedLabel) +
    RECEIVED_COLOR(receivedSerializedTrimmed)
  );
};
