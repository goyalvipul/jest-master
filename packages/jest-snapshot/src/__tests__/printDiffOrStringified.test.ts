/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import ansiRegex from 'ansi-regex';
import style from 'ansi-styles';
import {printDiffOrStringified} from '../print';
import {serialize, unescape} from '../utils';

// This is an experiment to read snapshots more easily:
// * to avoid first line misaligned because of opening double quote mark,
//   return string without calling print function to serialize it,
//   which also reduces extra escape sequences which is a subject of the tests!
// * to align lines, return <d> <g> <r> <y> tags which have same width
// * to see inline markup, return matching <i> and </i> tags
// * to see unexpected escape codes, do not return empty string as default

const convertStyles = (val: string): string =>
  val.replace(ansiRegex(), match => {
    switch (match) {
      case style.inverse.open:
        return '<i>';
      case style.inverse.close:
        return '</i>';

      case style.dim.open:
        return '<d>';
      case style.green.open:
        return '<g>';
      case style.red.open:
        return '<r>';
      case style.yellow.open:
        return '<y>';

      case style.dim.close:
      case style.green.close:
      case style.red.close:
      case style.yellow.close:
        return '</>';

      default:
        return match;
    }
  });

expect.addSnapshotSerializer({
  serialize(val: string): string {
    return val;
  },
  test(val: any): val is string {
    return typeof val === 'string';
  },
});

const testWithSerialize = (
  expected: any,
  received: any,
  expand: boolean,
): string => {
  // Simulate serializing the expected value as a snapshot,
  // and then returning actual and expected when match function fails.
  // Assume that the caller of printDiffOrStringified trims the strings.
  const expectedSerializedTrimmed = unescape(serialize(expected)).trim();
  const receivedSerializedTrimmed = unescape(serialize(received)).trim();

  return convertStyles(
    printDiffOrStringified(
      expectedSerializedTrimmed,
      receivedSerializedTrimmed,
      received,
      'Snapshot',
      'Received',
      expand,
    ),
  );
};

const testWithoutSerialize = (
  expected: string,
  received: string,
  expand: boolean,
): string =>
  convertStyles(
    printDiffOrStringified(
      expected,
      received,
      received,
      'Snapshot',
      'Received',
      expand,
    ),
  );

describe('backtick', () => {
  test('single line expected and received', () => {
    const expected = 'var foo = `backtick`;';
    const received = 'var foo = tag`backtick`;';

    expect(testWithSerialize(expected, received, false)).toMatchSnapshot();
  });
});

describe('empty string', () => {
  test('expected and received single line', () => {
    const expected = '';
    const received = 'single line string';

    expect(testWithSerialize(expected, received, false)).toMatchSnapshot();
  });

  test('received and expected multi line', () => {
    const expected = 'multi\nline\nstring';
    const received = '';

    expect(testWithSerialize(expected, received, false)).toMatchSnapshot();
  });
});

describe('escape', () => {
  test('double quote marks in string', () => {
    const expected = 'What does "oobleck" mean?';
    const received = 'What does "ewbleck" mean?';

    expect(testWithSerialize(expected, received, false)).toMatchSnapshot();
  });

  test('backslash in multi line string', () => {
    const expected = 'Forward / slash and back \\ slash';
    const received = 'Forward / slash\nBack \\ slash';

    expect(testWithSerialize(expected, received, false)).toMatchSnapshot();
  });

  test('backslash in single line string', () => {
    const expected = 'forward / slash and back \\ slash';
    const received = 'Forward / slash and back \\ slash';

    expect(testWithSerialize(expected, received, false)).toMatchSnapshot();
  });

  test('regexp', () => {
    const expected = /\\(")/g;
    const received = /\\(")/;

    expect(testWithSerialize(expected, received, false)).toMatchSnapshot();
  });
});

describe('expand', () => {
  // prettier/pull/5272
  const expected = [
    'type TypeName<T> =',
    'T extends string ? "string" :',
    'T extends number ? "number" :',
    'T extends boolean ? "boolean" :',
    'T extends undefined ? "undefined" :',
    'T extends Function ? "function" :',
    '"object";',
    '~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~',
    'type TypeName<T> = T extends string',
    '? "string"',
    ': T extends number',
    '? "number"',
    ': T extends boolean',
    '? "boolean"',
    ': T extends undefined',
    '? "undefined"',
    ': T extends Function ? "function" : "object";',
    '',
  ].join('\n');
  const received = [
    'type TypeName<T> =',
    'T extends string ? "string" :',
    'T extends number ? "number" :',
    'T extends boolean ? "boolean" :',
    'T extends undefined ? "undefined" :',
    'T extends Function ? "function" :',
    '"object";',
    '~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~',
    'type TypeName<T> = T extends string',
    '? "string"',
    ': T extends number',
    '? "number"',
    ': T extends boolean',
    '? "boolean"',
    ': T extends undefined',
    '? "undefined"',
    ': T extends Function',
    '? "function"',
    ': "object";',
    '',
  ].join('\n');

  test('false', () => {
    expect(testWithSerialize(expected, received, false)).toMatchSnapshot();
  });

  test('true', () => {
    expect(testWithSerialize(expected, received, true)).toMatchSnapshot();
  });
});

test('fallback to line diff', () => {
  const expected = [
    '[...a, ...b,];',
    '[...a, ...b];',
    '~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~',
    '[...a, ...b];',
    '[...a, ...b];',
    '',
  ].join('\n');
  const received = [
    '====================================options=====================================',
    'parsers: ["flow", "typescript"]',
    'printWidth: 80',
    '                                                                                | printWidth',
    '=====================================input======================================',
    '[...a, ...b,];',
    '[...a, ...b];',
    '',
    '=====================================output=====================================',
    '[...a, ...b];',
    '[...a, ...b];',
    '',
    '================================================================================',
  ].join('\n');

  expect(testWithSerialize(expected, received, false)).toMatchSnapshot();
});

describe('isLineDiffable', () => {
  describe('false', () => {
    test('boolean', () => {
      const expected = true;
      const received = false;

      expect(testWithSerialize(expected, received, false)).toMatchSnapshot();
    });

    test('number', () => {
      const expected = -0;
      const received = NaN;

      expect(testWithSerialize(expected, received, false)).toMatchSnapshot();
    });
  });

  describe('true', () => {
    test('array', () => {
      const expected0 = {
        code: 4011,
        weight: 2.13,
      };
      const expected1 = {
        code: 4019,
        count: 4,
      };

      const expected = [expected0, expected1];
      const received = [
        {_id: 'b14680dec683e744ada1f2fe08614086', ...expected0},
        {_id: '7fc63ff01769c4fa7d9279e97e307829', ...expected1},
      ];

      expect(testWithSerialize(expected, received, false)).toMatchSnapshot();
    });

    test('object', () => {
      const type = 'img';
      const expected = {
        props: {
          className: 'logo',
          src: '/img/jest.png',
        },
        type,
      };
      const received = {
        props: {
          alt: 'Jest logo',
          class: 'logo',
          src: '/img/jest.svg',
        },
        type,
      };

      expect(testWithSerialize(expected, received, false)).toMatchSnapshot();
    });

    test('single line expected and received', () => {
      const expected = [];
      const received = {};

      expect(testWithSerialize(expected, received, false)).toMatchSnapshot();
    });

    test('single line expected and multi line received', () => {
      const expected = [];
      const received = [0];

      expect(testWithSerialize(expected, received, false)).toMatchSnapshot();
    });
  });
});

test('multi line small change in one line and other is unchanged', () => {
  const expected =
    "There is no route defined for key 'Settings'.\nMust be one of: 'Home'";
  const received =
    "There is no route defined for key Settings.\nMust be one of: 'Home'";

  expect(testWithSerialize(expected, received, false)).toMatchSnapshot();
});

test('multi line small changes', () => {
  const expected = [
    '    69 | ',
    "    70 | test('assert.doesNotThrow', () => {",
    '  > 71 |   assert.doesNotThrow(() => {',
    '       |          ^',
    "    72 |     throw Error('err!');",
    '    73 |   });',
    '    74 | });',
    '    at Object.doesNotThrow (__tests__/assertionError.test.js:71:10)',
  ].join('\n');
  const received = [
    '    68 | ',
    "    69 | test('assert.doesNotThrow', () => {",
    '  > 70 |   assert.doesNotThrow(() => {',
    '       |          ^',
    "    71 |     throw Error('err!');",
    '    72 |   });',
    '    73 | });',
    '    at Object.doesNotThrow (__tests__/assertionError.test.js:70:10)',
  ].join('\n');

  expect(testWithSerialize(expected, received, false)).toMatchSnapshot();
});

test('single line large changes', () => {
  const expected = 'Array length must be a finite positive integer';
  const received = 'Invalid array length';

  expect(testWithSerialize(expected, received, false)).toMatchSnapshot();
});

describe('without serialize', () => {
  test('backtick single line expected and received', () => {
    const expected = 'var foo = `backtick`;';
    const received = 'var foo = `back${x}tick`;';

    expect(testWithoutSerialize(expected, received, false)).toMatchSnapshot();
  });

  test('backtick single line expected and multi line received', () => {
    const expected = 'var foo = `backtick`;';
    const received = 'var foo = `back\ntick`;';

    expect(testWithoutSerialize(expected, received, false)).toMatchSnapshot();
  });

  test('prettier/pull/5590', () => {
    const expected = [
      '====================================options=====================================',
      'parsers: ["html"]',
      'printWidth: 80',
      '                                                                                | printWidth',
      '=====================================input======================================',
      `<img src="test.png" alt='John "ShotGun" Nelson'>`,
      '',
      '=====================================output=====================================',
      '<img src="test.png" alt="John &quot;ShotGun&quot; Nelson" />',
      '',
      '================================================================================',
    ].join('\n');
    const received = [
      '====================================options=====================================',
      'parsers: ["html"]',
      'printWidth: 80',
      '                                                                                | printWidth',
      '=====================================input======================================',
      `<img src="test.png" alt='John "ShotGun" Nelson'>`,
      '',
      '=====================================output=====================================',
      `<img src="test.png" alt='John "ShotGun" Nelson' />`,
      '',
      '================================================================================',
    ].join('\n');

    expect(testWithoutSerialize(expected, received, false)).toMatchSnapshot();
  });
});
