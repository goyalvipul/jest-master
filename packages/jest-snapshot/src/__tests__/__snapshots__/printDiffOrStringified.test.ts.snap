// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`backtick single line expected and received 1`] = `
Snapshot: <g>"var foo = \`backtick\`;"</>
Received: <r>"var foo = <i>tag</i>\`backtick\`;"</>
`;

exports[`empty string expected and received single line 1`] = `
Snapshot: <g>""</>
Received: <r>"single line string"</>
`;

exports[`empty string received and expected multi line 1`] = `
Snapshot: <g>"multi</>
<g>line</>
<g>string"</>
Received: <r>""</>
`;

exports[`escape backslash in multi line string 1`] = `
<g>- Snapshot</>
<r>+ Received</>

<g>- Forward / slash<i> and b</i>ack \\ slash</>
<r>+ Forward / slash</>
<r>+ <i>B</i>ack \\ slash</>
`;

exports[`escape backslash in single line string 1`] = `
Snapshot: <g>"<i>f</i>orward / slash and back \\\\ slash"</>
Received: <r>"<i>F</i>orward / slash and back \\\\ slash"</>
`;

exports[`escape double quote marks in string 1`] = `
Snapshot: <g>"What does \\"<i>oo</i>bleck\\" mean?"</>
Received: <r>"What does \\"<i>ew</i>bleck\\" mean?"</>
`;

exports[`escape regexp 1`] = `
Snapshot: <g>/\\\\\\\\\\("\\)/g</>
Received: <r>/\\\\\\\\\\("\\)/</>
`;

exports[`expand false 1`] = `
<g>- Snapshot</>
<r>+ Received</>

<y>@@ -12,7 +12,9 @@</>
<d>  ? "number"</>
<d>  : T extends boolean</>
<d>  ? "boolean"</>
<d>  : T extends undefined</>
<d>  ? "undefined"</>
<g>- : T extends Function<i> </i>? "function"<i> </i>: "object";</>
<r>+ : T extends Function</>
<r>+ ? "function"</>
<r>+ : "object";</>
<d>  ↵</>
`;

exports[`expand true 1`] = `
<g>- Snapshot</>
<r>+ Received</>

<d>  type TypeName<T> =</>
<d>  T extends string ? "string" :</>
<d>  T extends number ? "number" :</>
<d>  T extends boolean ? "boolean" :</>
<d>  T extends undefined ? "undefined" :</>
<d>  T extends Function ? "function" :</>
<d>  "object";</>
<d>  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</>
<d>  type TypeName<T> = T extends string</>
<d>  ? "string"</>
<d>  : T extends number</>
<d>  ? "number"</>
<d>  : T extends boolean</>
<d>  ? "boolean"</>
<d>  : T extends undefined</>
<d>  ? "undefined"</>
<g>- : T extends Function<i> </i>? "function"<i> </i>: "object";</>
<r>+ : T extends Function</>
<r>+ ? "function"</>
<r>+ : "object";</>
<d>  ↵</>
`;

exports[`fallback to line diff 1`] = `
<g>- Snapshot</>
<r>+ Received</>

<r>+ ====================================options=====================================</>
<r>+ parsers: ["flow", "typescript"]</>
<r>+ printWidth: 80</>
<r>+                                                                                 | printWidth</>
<r>+ =====================================input======================================</>
<d>  [...a, ...b,];</>
<d>  [...a, ...b];</>
<g>- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</>
<r>+ </>
<r>+ =====================================output=====================================</>
<d>  [...a, ...b];</>
<d>  [...a, ...b];</>
<d>  </>
<r>+ ================================================================================</>
`;

exports[`isLineDiffable false boolean 1`] = `
Snapshot: <g>true</>
Received: <r>false</>
`;

exports[`isLineDiffable false number 1`] = `
Snapshot: <g>-0</>
Received: <r>NaN</>
`;

exports[`isLineDiffable true array 1`] = `
<g>- Snapshot</>
<r>+ Received</>

<d>  Array [</>
<d>    Object {</>
<r>+     "_id": "b14680dec683e744ada1f2fe08614086",</>
<d>      "code": 4011,</>
<d>      "weight": 2.13,</>
<d>    },</>
<d>    Object {</>
<r>+     "_id": "7fc63ff01769c4fa7d9279e97e307829",</>
<d>      "code": 4019,</>
<d>      "count": 4,</>
<d>    },</>
<d>  ]</>
`;

exports[`isLineDiffable true object 1`] = `
<g>- Snapshot</>
<r>+ Received</>

<d>  Object {</>
<d>    "props": Object {</>
<g>-     "className": "logo",</>
<g>-     "src": "/img/jest.png",</>
<r>+     "alt": "Jest logo",</>
<r>+     "class": "logo",</>
<r>+     "src": "/img/jest.svg",</>
<d>    },</>
<d>    "type": "img",</>
<d>  }</>
`;

exports[`isLineDiffable true single line expected and multi line received 1`] = `
<g>- Snapshot</>
<r>+ Received</>

<g>- Array []</>
<r>+ Array [</>
<r>+   0,</>
<r>+ ]</>
`;

exports[`isLineDiffable true single line expected and received 1`] = `
Snapshot: <g>Array []</>
Received: <r>Object {}</>
`;

exports[`multi line small change in one line and other is unchanged 1`] = `
<g>- Snapshot</>
<r>+ Received</>

<g>- There is no route defined for key <i>'</i>Settings<i>'</i>.</>
<r>+ There is no route defined for key Settings.</>
<d>  Must be one of: 'Home'</>
`;

exports[`multi line small changes 1`] = `
<g>- Snapshot</>
<r>+ Received</>

<g>-     6<i>9</i> |·</>
<r>+     6<i>8</i> |·</>
<g>-     <i>70</i> | test('assert.doesNotThrow', () => {</>
<r>+     <i>69</i> | test('assert.doesNotThrow', () => {</>
<g>-   > 7<i>1</i> |   assert.doesNotThrow(() => {</>
<r>+   > 7<i>0</i> |   assert.doesNotThrow(() => {</>
<d>         |          ^</>
<g>-     7<i>2</i> |     throw Error('err!');</>
<r>+     7<i>1</i> |     throw Error('err!');</>
<g>-     7<i>3</i> |   });</>
<r>+     7<i>2</i> |   });</>
<g>-     7<i>4</i> | });</>
<r>+     7<i>3</i> | });</>
<g>-     at Object.doesNotThrow (__tests__/assertionError.test.js:7<i>1</i>:10)</>
<r>+     at Object.doesNotThrow (__tests__/assertionError.test.js:7<i>0</i>:10)</>
`;

exports[`single line large changes 1`] = `
Snapshot: <g>"<i>A</i>rray length<i> must be a finite positive integer</i>"</>
Received: <r>"<i>Invalid a</i>rray length"</>
`;

exports[`without serialize backtick single line expected and multi line received 1`] = `
<g>- Snapshot</>
<r>+ Received</>

<g>- var foo = \`backtick\`;</>
<r>+ var foo = \`back</>
<r>+ tick\`;</>
`;

exports[`without serialize backtick single line expected and received 1`] = `
Snapshot: <g>var foo = \`backtick\`;</>
Received: <r>var foo = \`back<i>\${x}</i>tick\`;</>
`;

exports[`without serialize prettier/pull/5590 1`] = `
<g>- Snapshot</>
<r>+ Received</>

<y>@@ -4,8 +4,8 @@</>
<d>                                                                                  | printWidth</>
<d>  =====================================input======================================</>
<d>  <img src="test.png" alt='John "ShotGun" Nelson'></>

<d>  =====================================output=====================================</>
<g>- <img src="test.png" alt=<i>"</i>John <i>&quot;</i>ShotGun<i>&quot;</i> Nelson<i>"</i> /></>
<r>+ <img src="test.png" alt=<i>'</i>John <i>"</i>ShotGun<i>"</i> Nelson<i>'</i> /></>

<d>  ================================================================================</>
`;
